### PR TITLE
Refactor parsing and expose new WASM interface

### PR DESCRIPTION
**Pull request by adambabik**
Highlights:
* Parsing is simplified and refactored to handle edge cases more easily like nested fenced code blocks.
* Rendering process exposes each block separately so that when using the Notebook interface, we have blocks corresponding to cells.

To test the new WASM interface, run:

```
make wasm
cd examples/web
python -m http.server 9000
```

And open the address in a browser. You should see the new output. Check out `examples/web/index.html` to learn more about the new global functions.

## Todo

- [x] Go through these MDs:
   * [DevOps Handbook](https://github.com/freeCodeCamp/freeCodeCamp/blob/084bf7600f2f8fcdcb2bb82abe2f842d457b270a/docs/devops.md)
- [x] Implement the case when editing a block results in its split. For example, a block containing a single paragraph might be edited and converted into two paragraphs which result in two blocks each containing a single paragraph.

Closes https://github.com/stateful/runme/issues/49 https://github.com/stateful/runme/issues/50

### CODE REVIEW

**Comment by christian-bromann** [comment]
Another markdown example which we can use to test this fix: https://github.com/vercel/next.js/blob/canary/contributing/core/developing.md

**Comment by adambabik** [comment]
@christian-bromann thanks for this example. It helped me to find a bug which was fixed in https://github.com/stateful/runme/pull/53/commits/af65f91952e567209e836202f399f3998684ebfc.

**Review by sourishkrout — CHANGES_REQUESTED** [review]
1. Consider adding the streamlined WASM interface.
2. Marshaling runs into deadlock with this file https://gist.github.com/sourishkrout/4f6bc408bd2287099fac39e2b99a5f9a - circular reference?

@adambabik please lemme know if you have questions

File: internal/document/block.go [hunk 1]
```diff
@@ -205,67 +249,210 @@ func (b *CodeBlock) Name() string {
...
+				start++
+			}
+			for ch := b.source[start]; ch == '\n'; ch = b.source[start] {
+				start++
+			}
+
+			nextBlock := b.inner.NextSibling()
+			for nextBlock != nil && !hasLine(nextBlock) {
+				nextBlock.NextSibling()
+			}
```
Comment (sourishkrout) [inline internal/document/block.go:388]: Never mind the ciruclar ref comment... this is where the infinite loop happens @adambabik

File: internal/document/block.go [hunk 2]
```diff
@@ -205,67 +249,210 @@ func (b *CodeBlock) Name() string {
...
+				_, start = getContentRange(b.source, previousBlock)
+				start++
+			}
+			for ch := b.source[start]; ch == '\n'; ch = b.source[start] {
+				start++
+			}
+
+			nextBlock := b.inner.NextSibling()
+			for nextBlock != nil && !hasLine(nextBlock) {
+				nextBlock.NextSibling()
```
Comment (sourishkrout) [inline internal/document/block.go:387]:
```suggestion
				nextBlock = nextBlock.NextSibling()
```
i believe this will fix it

File: internal/document/block.go [hunk 3]
```diff
@@ -205,67 +249,210 @@ func (b *CodeBlock) Name() string {
...
+				stop = stop + idx
+			}
+
+			b.content = string(b.source[start:stop])
+			b.start = start
+			b.stop = stop
+		case ast.KindThematicBreak:
+			previousBlock := b.inner.PreviousSibling()
+			for previousBlock != nil && !hasLine(previousBlock) {
+				previousBlock.PreviousSibling()
```
Comment (sourishkrout) [inline internal/document/block.go:373]:
```suggestion
				previousBlock = previousBlock.PreviousSibling()
```
and here

File: web/main_js.go [hunk 4]
```diff
@@ -18,51 +20,93 @@ var (
...
+	result := make(map[string]interface{})
+	_ = json.Unmarshal(b.Bytes(), &result)
+	return result
+}
+
+func getSource(this js.Value, args []js.Value) any {
+	if parsed == nil {
+		return errors.New("call initialize() first")
+	}
+	return parsed.Source()
```
Comment (sourishkrout) [inline web/main_js.go:87]: if you try calling this from the browser @adambabik, it'll throw an error. Suspecting it does know how what to do what to do with a byte array.

File: web/main_js.go [hunk 5]
```diff
@@ -18,51 +20,93 @@ var (
...
+	if parsed == nil {
+		return errors.New("call initialize() first")
+	}
+
+	idx := args[0].Int()
+	newSource := args[1].String()
+
+	updater := document.NewUpdater(parsed)
+	if err := updater.UpdateBlock(idx, newSource); err != nil {
+		return err
```
Comment (sourishkrout) [inline web/main_js.go:100]: same with @adambabik, it appears it can't handle the Golang Error type

**Comment by sourishkrout** [comment]
question for you @adambabik: looks like `source: '```\nnode ./scripts/stdin.js\n```'` always includes the triple backticks of fenced code blocks now. will this be universally true for `type=code`?

**Comment by adambabik** [comment]
> question for you @adambabik: looks like `` source: '```\nnode ./scripts/stdin.js\n```' `` always includes the opening/closing triple backticks of fenced code blocks now. will this be universally true for `type=code`?

Not sure what you mean by `type=code`. If you mean also code blocks (not fenced) then no. These are called fence marks and only exists for the fenced code blocks and we only recognize those as executable.

I can get rid of the fence marks if that's problematic. VS Code interface for code cells probably accepts raw code without them. Using `lines` would be worse because they strip out all prefix and trailing white spaces and we want to keep them - this is an existing flaw where code blocks look like a wall of text without indentations.

File: web/main_js.go (hunk 4 above)
Comment (adambabik) [reply web/main_js.go]: Good catch, I didn't verify in the `example.html` this function. It's fixed and it works now in the browser.

**Comment by sourishkrout** [comment]
> Not sure what you mean by `type=code`. If you mean also code blocks (not fenced) then no. These are called fence marks and only exists for the fenced code blocks and we only recognize those as executable.
> 
> I can get rid of the fence marks if that's problematic. VS Code interface for code cells probably accepts raw code without them. Using `lines` would be worse because they strip out all prefix and trailing white spaces and we want to keep them - this is an existing flaw where code blocks look like a wall of text without indentations.

Actually for visualization keeping the backticks is better, since when we have unsupported languages (non-shell-based) we do want to render them as the original markdown.

I think it makes sense to stay as close to the source of truth as possible. They only place it came up was with the language detection model. If I leave the backticks in everything will be guessed as markdown. I just created a purpose specific normalization function.

**Comment by sourishkrout** [comment]
Integration available here: https://github.com/stateful/vscode-runme/pull/52

File: examples/web/index.html [hunk 6]
```diff
@@ -18,11 +18,28 @@
...
       const go = new Go();
-      WebAssembly.instantiateStreaming(fetch("runme.wasm"), go.importObject).then((runme) => {
-        go.run(runme.instance);
-        console.log(GetSnippets(markdown));
-        console.log(GetDocument(markdown));
-        console.log(PrepareScript(["bash -c 'echo \"1\"'"]));
+      WebAssembly.instantiateStreaming(fetch("runme.wasm"), go.importObject).then((obj) => {
+        go.run(obj.instance);
+
+        console.log("return error", Runme.getSource());
```
Comment (sourishkrout) [inline examples/web/index.html:24]: Technically it's odd to "return" errors in JS, but let's not get hung up on it. Lets plan to iterate on errors as part of building out the editing stuff.

**Review by sourishkrout — APPROVED** [review]
Ready to merge @adambabik?

File: examples/web/index.html (hunk 6 above)
Comment (adambabik) [reply examples/web/index.html:24]: There is no other way AFAIK. I could return a string but I guess it makes sense - an exported JS function returns a JS `Error` instance. Or I could return `null` :) We will get back to it.
